llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -5411,6 +5411,14 @@ SITargetLowering::EmitInstrWithCustomInserter(MachineInstr &MI,
     MI.eraseFromParent();
     return BB;
   }
+  case AMDGPU::S_BARRIER_SIGNAL_ISFIRST_IMM: {
+    // Set SCC to true, in case the barrier instruction gets converted to a NOP.
+    BuildMI(*BB, MI.getIterator(), MI.getDebugLoc(),
+            TII->get(AMDGPU::S_CMP_EQ_U32))
+        .addImm(0)
+        .addImm(0);
```

Comment:
Can this leave a dead instruction behind if it doesn't?

Reply:
Perhaps the comment isn't clear enough. The point is that *hardware* at runtime can convert the barrier instructions into NOPs.

I suppose we could remove the barrier instruction ourselves if the kernel size is statically known to be too small, but I would we hope we'd do that earlier in LLVM IR since it might expose more optimization opportunities.

Reply:
So it's sort of a hardware bug? The ordinary behavior is SCC is always set to 1? And if it nops it leaves it in the previous state?